libmscore/changeMap.cpp

```diff
@@ -0,0 +1,487 @@
+      eventIter--;
+      bool foundRamp = false;
+      ChangeEvent& rampFound = eventIter.value();       // only used to init
+      for (auto& event : values(eventIter.key())) {
+            if (event.type == ChangeEventType::RAMP) {
+                  foundRamp = true;
+                  rampFound = event;
+                  }
+            }
+
+      if (!foundRamp) {
+            // Last event must be a fix, since there are max two events at one tick
+            return eventIter.value().value;
+            }
+
+      if (tick >= rampFound.etick) {
+            return rampFound.cachedEnd;
+            }
+      else {
+            // Do some maths!
+            return interpolate(eventIter.value(), tick);
+            }
+      }
+
+//---------------------------------------------------------
+//   addFixed
+//---------------------------------------------------------
+
+void ChangeMap::addFixed(Fraction tick, int value)
+      {
+      insert(tick, ChangeEvent(tick, value));
+      cleanedUp = false;
+      }
+
+//---------------------------------------------------------
+//   addRamp
+///   A `change` of 0 means that the change in velocity should be calculated from the next fixed
+///   velocity event.
+//---------------------------------------------------------
+
+void ChangeMap::addRamp(Fraction stick, Fraction etick, int change, ChangeMethod method, ChangeDirection direction)
+      {
+      change = abs(change);
+      change *= (direction == ChangeDirection::INCREASING) ? 1 : -1;
+      insert(stick, ChangeEvent(stick, etick, change, method, direction));
+      cleanedUp = false;
+      }
+
+//---------------------------------------------------------
+//   cleanup
+//---------------------------------------------------------
+
+void ChangeMap::cleanup()
+      {
+      //
+      // Stage 0: put the ramps in size order if they start at the same point
+      //
+      for (auto& tick : uniqueKeys()) {
+            // `events` will contain all the ramps at this tick
+            std::vector<ChangeEvent> events;
+            for (auto& event : values(tick)) {
+                  if (event.type == ChangeEventType::FIX)
+                        continue;
+                  events.push_back(event);
+                  }
+
+            if (int(events.size()) > 1) {
+                  // new order is a map of the NEGATIVE length of a ramp to a copy of its event
+                  QMultiMap<Fraction, ChangeEvent> newOrder;
+                  for (auto& event : events) {
+                        newOrder.insert(event.tick - event.etick, event);
+                        // And clear out the events at this tick
+                        remove(tick, event);
+                        }
+
+                  // newOrder's values are implicitally ordered by the key, which is the
+                  // NEGATIVE length of the ramp.
+                  for (auto& event : newOrder.values()) {
+                        insert(tick, event);
+                        }
```

Comment:
If I am correct, all this `newOrder` procedure is to get `RAMP` events sorted by their lengths, isn't it? Why not to use `std::sort` on `events` then? (I see, maybe to prevent subtracting `etick` and `tick` multiple times. Then, again, storing length instead of absolute tick values would naturally solve this issue.)

Reply:
Fixed by next commit